client/hotrod-client/src/main/java/org/infinispan/client/hotrod/telemetry/impl/TelemetryServiceFactory.java

```diff
@@ -0,0 +1,53 @@
+package org.infinispan.client.hotrod.telemetry.impl;
+
+import org.infinispan.client.hotrod.logging.Log;
+import org.infinispan.client.hotrod.logging.LogFactory;
+
+public class TelemetryServiceFactory {
+
+   private static final Log log = LogFactory.getLog(TelemetryService.class, Log.class);
+
+   public static final TelemetryServiceFactory INSTANCE = new TelemetryServiceFactory();
+
+   private volatile TelemetryService telemetryService;
+   private volatile boolean configured = false;
+   private volatile boolean loggedDisabledTracing = false;
+
+   private TelemetryServiceFactory() {
+   }
+
+   public TelemetryService telemetryService(boolean tracingPropagationEnabled) {
+      if (!tracingPropagationEnabled) {
+         if (loggedDisabledTracing == false) {
+            loggedDisabledTracing = true;
+            log.openTelemetryPropagationDisabled();
+         }
+         return null;
+      }
+
+      return telemetryService();
+   }
+
+   private synchronized TelemetryService telemetryService() {
+      if (configured) {
+         return telemetryService;
+      }
+
+      try {
+         Class.forName("io.opentelemetry.api.trace.propagation.W3CTraceContextPropagator", false,
+               TelemetryService.class.getClassLoader());
+      } catch (ClassNotFoundException e) {
+         log.noOpenTelemetryAPI();
```

Comment:
This try catch block (log including) is going to be repeated regularly due to `tracingPropagationEnabled` being true by default. OpenTelemetry API is not on a classpath by default though.

While having tracingPropagationEnabled se to true by default is not an issue as user can just add a dependency to their projects wouldn't it be better to attempt to load OpenTelemetry API only once and effectively log whether it's available only once?

Reply:
Right. I'm fixing...

Reply:
now it should work I suppose.... thanks @Crumby

Reply:
You should move the `Class.forName()` stuff inside the `TelemetryServiceFactory` constructor so it is verified only once.
Then make the `TelemetryService telemetryService(boolean tracingPropagationEnabled)` method static (and use the `INSTANCE` field in there). No need for volatiles, synchronized or some such.

Reply:
I think that `Class.forName()` can be invoked at least once also in the current solution, since with the last change we made by the time we execute `Class.forName` the `configured` has been set to true so that the same line of code won't be accessible next time that the method is invoked.

Anyway I agree with you that the solution looks too complex. The reason I didn’t put the build into the constructor is simply because I wanted to avoid the `Class.forName()` execution in case tracingPropagationEnabled will be false for all created clients.

I think we can avoid to not have this, since the default for tracingPropagationEnabled is true and the feature is very advanced, so 99.9999% time the value will be true, so we need to execute `Class.forName()` in any case.

Going to simplify it, applying your idea. @tristantarrant